pkg/storegateway/series_refs.go

```diff
@@ -830,69 +832,80 @@ func metasToChunks(blockID ulid.ULID, metas []chunks.Meta) []seriesChunkRef {
-	if err := decodeSnappyGob(data, &entry); err != nil {
+	data, err := snappy.Decode(nil, data)
+	if err != nil {
+		logSeriesForPostingsCacheEvent(ctx, logger, userID, blockID, matchers, shard, postingsKey, "msg", "can't decode series cache snappy", "err", err)
+		return seriesChunkRefsSet{}, false
+	}
+	var entry storepb.CachedSeries
+	if err = entry.Unmarshal(data); err != nil {
 		logSeriesForPostingsCacheEvent(ctx, logger, userID, blockID, matchers, shard, postingsKey, "msg", "can't decode series cache", "err", err)
 		return seriesChunkRefsSet{}, false
 	}
-	if entry.MatchersKey != matchersKey {
+
+	if entry.MatchersKey != string(matchersKey) {
 		logSeriesForPostingsCacheEvent(ctx, logger, userID, blockID, matchers, shard, postingsKey, "msg", "cached series entry key doesn't match, possible collision", "cached_matchers", entry.MatchersKey)
 		return seriesChunkRefsSet{}, false
 	}
-	if entry.Shard != maybeNilShard(shard) {
-		logSeriesForPostingsCacheEvent(ctx, logger, userID, blockID, matchers, shard, postingsKey, "msg", "cached series shard doesn't match, possible collision", "cached_shard", entry.Shard)
+	if nonNilShard := maybeNilShard(shard); entry.ShardIndex != nonNilShard.ShardIndex || entry.ShardCount != nonNilShard.ShardCount {
+		logSeriesForPostingsCacheEvent(ctx, logger, userID, blockID, matchers, shard, postingsKey, "msg", "cached series shard doesn't match, possible collision", "cached_shard_index", entry.ShardIndex, "cached_shard_count", entry.ShardCount)
 		return seriesChunkRefsSet{}, false
 	}
 
 	// This can be released by the caller because loadingSeriesChunkRefsSetIterator (where this function is called) doesn't retain it
 	// after Next() will be called again.
-	res := newSeriesChunkRefsSet(len(entry.LabelSets), true)
-	for _, lset := range entry.LabelSets {
+	res := newSeriesChunkRefsSet(len(entry.Series), true)
+	for _, lset := range entry.Series {
 		res.series = append(res.series, seriesChunkRefs{
-			lset: lset,
+			lset: mimirpb.FromLabelAdaptersToLabels(lset.Labels),
 		})
 	}
 	return res, true
 }
 
 func storeCachedSeriesForPostings(ctx context.Context, indexCache indexcache.IndexCache, userID string, blockID ulid.ULID, matchers []*labels.Matcher, shard *sharding.ShardSelector, postingsKey indexcache.PostingsKey, set seriesChunkRefsSet, logger log.Logger) {
-	entry := seriesForPostingsCacheEntry{
-		LabelSets:   make([]labels.Labels, set.len()),
-		MatchersKey: indexcache.CanonicalLabelMatchersKey(matchers),
-		Shard:       maybeNilShard(shard),
+	nonNilShard := maybeNilShard(shard)
+	matchersKey := indexcache.CanonicalLabelMatchersKey(matchers)
+	data, err := encodeCachedSeriesForPostings(set, matchersKey, nonNilShard)
+	if err != nil {
+		logSeriesForPostingsCacheEvent(ctx, logger, userID, blockID, matchers, shard, postingsKey, "msg", "can't encode series for caching", "err", err)
+		return
+	}
+	indexCache.StoreSeriesForPostings(ctx, userID, blockID, matchersKey, shard, postingsKey, data)
```

Comment:
Since we're changing the format of the data, we would need to change the memcache key for this cache.

Reply:
i don't think we need to. the cache entry will be found as invalid and populated again.

Reply:
That's like exception-driven development and not the best practice IMO.

Reply:
i don't have a strong preference. But i think handling bad data in the cache should be a feature, not a quirk. Would you change your mind if I added a test to make sure we handle invalid data from the cache?

Reply:
I think the validation feature should always distinguish between an expected data format update versus an unexpected issue (which should be logged and warn the operator).

This feature is still experimental and _most likely_ nobody uses it yet, but an upgrade that complains about lots of invalid cache entries is scary: an upgrade should go smooth and quiet.

Reply:
ok, i see what you mean 👍 

`StoreSeriesForPostings` already uses a different cache key than `StoreSeries`, so whenever someone starts using streaming they shouldn't see errors

I can change the cache key again

Reply:
I meant someone changing from previous `main` commit to this new commit :)

Reply:
I changed it to `SfP:`

Reply:
changed to SP2